Source/WebCore/rendering/RenderElement.cpp

```diff
@@ -2489,4 +2489,14 @@ void RenderElement::layoutIfNeeded()
     layout();
 }
 
+SingleThreadWeakPtr<RenderInline> RenderElement::writingSuggestionsRenderer() const
+{
+    return hasRareData() ? rareData().writingSuggestionsRenderer : nullptr;
+}
+
+void RenderElement::setWritingSuggestionsRenderer(RenderInline& renderer)
+{
+    ensureRareData().writingSuggestionsRenderer = renderer;
+}
```

Comment:
Does this need to be on RenderElement? (as opposed to a RenderBlock or even RenderBlockFlow?)

Reply:
I’m not sure; what’s the difference between those three types?

Reply:
Do subclasses of RenderElement that are not RenderBlock / RenderBlockFlow ever get writing suggestions? If not, then you can store the renderers on the subclasses.

Reply:
I think theoretically any text-like RenderElement can get them

Reply:
what does text-like RenderElement mean?

Reply:
I guess I meant like RenderBlock, RenderBlockFlow, RenderInline, and RenderText (I think, at least), and not stuff like RenderImage

Reply:
Since this is writing suggestion on text content I think it's safe to have it on RenderBlockFlow and also RenderInline is only valid in the context of RenderBlockFlow. (btw RenderText is not a RenderElement)

Reply:
ah ok, apologies. will fix!